microscope/lasers/sapphire.py

```diff
@@ -0,0 +1,211 @@
+#!/usr/bin/python
+# -*- coding: utf-8
+#
+# Copyright 2016 Mick Phillips (mick.phillips@gmail.com)
+# and 2017 Ian Dobbie (Ian.Dobbie@gmail.com)
+# and 2018 Tiago Susano Pinto (tiagosusanopinto@gmail.com)
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+import serial
+import threading
+import time
+from microscope import devices
+import functools
+
+
+def lock_comms(func):
+    """A decorator to flush the input buffer prior to issuing a command.
+
+    Locks the comms channel so that a function must finish all its comms
+    before another can run.
+    """
+    @functools.wraps(func)
+    def wrapper(self, *args, **kwargs):
+        with self.comms_lock:
+            return func(self, *args, **kwargs)
+
+    return wrapper
+
+
+class SapphireLaser(devices.LaserDevice):
+    def __init__(self, com=None, baud=19200, timeout=0.5, **kwargs):
+        # laser controller must run at 19200 baud, 8+1 bits,
+        # no parity or flow control
+        # timeout is recomended to be over 0.5
+        super(SapphireLaser, self).__init__()
+        self.connection = serial.Serial(port = com,
+            baudrate = baud, timeout = timeout,
+            stopbits = serial.STOPBITS_ONE,
+            bytesize = serial.EIGHTBITS, parity = serial.PARITY_NONE)
+        # Turning off command prompt
+        self.send('>=0')
+        # Head ID value is a float point value,
+        # but only the integer part is significant
+        headID = int(float(self.send('?hid')))
+        self._logger.info("Sapphire laser serial number: [%s]" % int(headID))
+
+        self.comms_lock = threading.RLock()
+
+    def _read(self, num_chars):
+        """Simple passthrough to read numChars from connection."""
+        return self.connection.read(num_chars).decode()
+
+    def _readline(self):
+        """Simple passthrough to read one line from connection."""
+        return self.connection.readline().strip().decode()
+
```

Comment:
The methods `_read` and `_readline` are implemented in the parent class and they don't decode the bytes. We shouldn't have this laser behave different from the others. If we do this change (and I don't think we should but let's follow that discussion on the PR, not inline), then it should be done on the parent class.